src/main/java/com/google/devtools/build/lib/skyframe/FilesystemValueChecker.java

```diff
@@ -488,31 +501,31 @@ private boolean actionValueIsDirtyWithDirectSystemCalls(
         actionValue.getAllTreeArtifactValues().entrySet()) {
       TreeArtifactValue tree = entry.getValue();
 
-      if (!tree.isEntirelyRemote()) {
-        for (Map.Entry<TreeFileArtifact, FileArtifactValue> childEntry :
-            tree.getChildValues().entrySet()) {
-          if (artifactIsDirtyWithDirectSystemCalls(
-              knownModifiedOutputFiles,
-              trustRemoteArtifacts,
-              childEntry,
-              modifiedOutputsReceiver)) {
-            isDirty = true;
-          }
+      for (Map.Entry<TreeFileArtifact, FileArtifactValue> childEntry :
```

Comment:
Did we previously have a bug here (by skipping the loop when `tree.isEntirelyRemote`)? Obviously, there is now a new condition (the TTL) that could cause us to consider remote metadata dirty. But wasn't that already possible before, e.g. if `trustRemoteArtifacts` was false?

Reply:
> Did we previously have a bug here (by skipping the loop when tree.isEntirelyRemote)?

No. If we were able to skip the loop, `trustRemoteArtifacts` has to be `true`.

However, previously, the assumption was if we have `RemoteFileArtifactValue`, there shouldn't be corresponding local file. Otherwise, the metadata should be `RegularFileArtifactValue`. So we do the optimization by skipping the local IOs if `tree.isEntirelyRemote`.

With recent changes, this assumption doesn't hold anymore. For example, we inject remote metadata first and then prefetch the file for local actions. But the metadata is still `RemoteFileArtifactValue` in skyframe. That being said, 
this does **not** introduce bugs. It's just performance hit for the next incremental build (likely invisible): when we have local file and remote metadata, the remote metadata is marked as dirty (because local files take preference) and we will eventually hit the local action cache and the `RemoteFileArtifactValue` will be replaced by `RegularFileArtifactValue` in skyframe.

Reply:
Anyway, we now should always check the tree file no matter `tree.isEntirelyRemote`.